src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs

```diff
@@ -14,44 +14,100 @@ namespace Microsoft.IdentityModel.Logging
+        /// <exception cref="ArgumentNullException">if <paramref name="toStringCallback"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="toStringCallbackUnsafe"/> is null.</exception>
+        /// <remarks>
+        /// Since even the payload may sometimes contain security artifacts, naïve disarm algorithms such as removing signatures
+        /// will not work. For now the <paramref name="toStringCallback"/> is currently unused. Future changes
+        /// may introduce a support for best effort disarm logging which will leverage <paramref name="toStringCallback"/>.
+        /// </remarks>
+        public SecurityArtifact(object argument, Func<object, string> toStringCallback, Func<object, string> toStringCallbackUnsafe)
+        {
+            if (toStringCallback == null)
+                throw new ArgumentNullException(nameof(toStringCallback));
+
+            if (toStringCallbackUnsafe == null)
+                throw new ArgumentNullException(nameof (toStringCallbackUnsafe));
+
+            Argument = argument;
+            _disarmCallback = toStringCallback;
+            _callbackUnsafe = toStringCallbackUnsafe;
+        }
+
+        /// <summary>
+        /// A dummy callback which can be leveraged to return a standard scrubbed token in the case where expected token is unknown.
+        /// </summary>
+        /// <param name="_">Ignored token.</param>
+        /// <returns>The standard scrubbed token string.</returns>
+        public static string UnknownSafeTokenCallback(object _)
+        {
+            return _scrubbedArtifact;
         }
 
         /// <summary>
         /// Returns a string that represents the <see cref="Argument"/>.
         /// </summary>
-        /// <returns><c>Null</c> if the <see cref="Argument"/> is <see langword="null"/>, otherwise calls <see cref="System.ValueType.ToString()"/> method of the <see cref="Argument"/>.</returns>
+        /// <returns><c>Null</c> if the <see cref="Argument"/> is <see langword="null"/>, otherwise calls the provided safe callback on <see cref="Argument"/>.</returns>
         public override string ToString()
         {
-            if (_callback != null && Argument != null)
-                return _callback(Argument);
-
-            return Argument?.ToString() ?? "null";
+            // Defense in depth, callers should never provide a null callback but publicly exposed method in LogHelper referencing this didn't check.
+            if (_disarmCallback == null)
+                return _scrubbedArtifact;
+            if (Argument == null)
+                return "null";
+            else
+                return _disarmCallback(Argument);
         }
 
         // <inheritdoc/>
         public string UnsafeToString()
         {
-            return Argument?.ToString() ?? "null";
+            if (_callbackUnsafe == null)
+                return ToString();
+            if (Argument == null)
+                return "null";
```

Comment:
do we cover the return "null" test case?

Reply:
Added a test for this case and refactored so that unsafe and safe callbacks use the same flow for this